Feature simbad query tap (#2856)
astropy/astroquery: docs/simbad/query_tap.rst

^^^^^^^^^^^^^^^^^^^^^^^^^^

The method `~astroquery.simbad.SimbadClass.query_tap` is called with a string containing the
ADQL query.

.. code-block:: SQL

  /*ADQL queries start with selecting the columns that will be in the output. Usually,
  the column name is sufficient. If there is a need to lift ambiguity, add the table
  name first (table_name.column_name). This is also where the number of rows is fixed
  (here 5).*/
  SELECT TOP 5 basic.ra, basic.dec, main_id, nbref
  /*Then comes the declaration of the tables to be included in the query. Here *basic* and
  *ids*. Their common column is *oid* in *basic* and *oidref* in *ident*.*/
  FROM basic JOIN ident ON basic.oid = ident.oidref
  /*The conditions come after. This query filters otypes that are not in any star
  category, specific redshifts, and the results should have an NGC name in their
  list of names.*/
  WHERE (otype != 'star..') AND (rvz_redshift < 1) AND (id LIKE 'NGC%')
  /*The result is then sorted so that the top 5 selected corresponds to
  the objects cited by the largest number of papers.*/
  ORDER BY nbref DESC

This ADQL query can be called with `~astroquery.simbad.SimbadClass.query_tap`: 

> .. a bit long due to the ordering

Reviewer: can we make it faster, e.g. the ordering is not really needed here if it makes it much slower?

Author: I could remove it, but I wanted that first query to present every TAP step (select, from, where, order by). The doctest output would also need to be skipped because the output will change very often. Is it a big issue that this one is skipped?

Reviewer: no, in this case it's OK to skip it then

Author: Actually, timing it with 

```python
import timeit
timeit.timeit("""Simbad.query_tap("SELECT TOP 5 basic.ra, basic.dec, main_id, nbref FROM basic JOIN ident ON basic.oid = ident.oidref WHERE ((otype != 'Cl*..') AND (rvz_redshift < 1) AND (id LIKE 'NGC%')) ORDER BY nbref DESC")""", number=1, setup="from astroquery.simbad import Simbad")
```

Gives 1.04s on my machine. It's slower than the average query (~0.15s) but not dramatic.

Reviewer: Oh, that's not slow at all, at least we have much bigger problems to clear out. 

As we don't run the remote test for PRs, I would indeed suggest to change the skip to an ignore output, so we would have a chance of picking up errors if/when it stops working.

.. doctest-remote-data::

    >>> from astroquery.simbad import Simbad
    >>> Simbad.find_columns_by_keyword("Radial velocity")
    <Table length=8>
      table_name    column_name   ...  unit                  ucd                
        object         object     ... object                object              
    ------------- --------------- ... ------ -----------------------------------
            basic     rvz_bibcode ...        meta.bib.bibcode;spect.dopplerVeloc
            basic         rvz_err ... km.s-1       stat.error;spect.dopplerVeloc
            basic    rvz_err_prec ...                                           
            basic        rvz_qual ...          meta.code.qual;spect.dopplerVeloc
            basic      rvz_radvel ... km.s-1              spect.dopplerVeloc.opt
            basic rvz_radvel_prec ...                                           
            basic        rvz_type ...                                           
    mesVelocities          origin ...                                  meta.note

Example TAP queries
^^^^^^^^^^^^^^^^^^^

This section lists more complex queries by looking at use cases from former astroquery issues.

> `Getting all bibcodes containing a certain type of measurement for a given object. <https://github.com/astropy/astroquery/issues/243>`__

Reviewer: No need to link to github issues, remove the link from the heading here and below, too

Author: Done :)

`~astroquery.simbad.SimbadClass.query_tap` (for Table Access Protocol) is the one

> query to rule them all. It allows to interrogate all the information in SIMBAD with the

Reviewer 2: ```suggestion
query to rule them all. It allows one to access all the information in SIMBAD with the
```
Minor English, but "Allow to <verb>" isn't an accepted construction.  Also, "interrogate" carries connotations that mean it's not so commonly used; it's technically acceptable, but I'd recommend "access" or "query" instead.

  the objects cited by the largest number of papers.*/
  ORDER BY nbref DESC

This ADQL query can be called with `~astroquery.simbad.SimbadClass.query_tap`: 

.. nbref changes often so we ignore the output here
.. doctest-remote-data::

    >>> from astroquery.simbad import Simbad
    >>> Simbad.query_tap("""SELECT TOP 5 basic.ra, basic.dec, main_id, nbref
    ...                  FROM basic JOIN ident ON basic.oid = ident.oidref
    ...                  WHERE (otype != 'Cl*..') AND (rvz_redshift < 1)
    ...                  AND (id LIKE 'NGC%')
    ...                  ORDER BY nbref DESC""") # doctest: +IGNORE_OUTPUT
        <Table length=5>
            ra                dec         main_id  nbref
           deg                deg                       
         float64            float64        object  int32
    ------------------ ------------------ -------- -----
    10.684708333333333 41.268750000000004    M  31 12412
    13.158333333333333 -72.80027777777778 NAME SMC 10875
       187.70593076725 12.391123246083334    M  87  7040
    148.96845833333333  69.67970277777778    M  82  5769
        23.46206906218 30.660175111980003    M  33  5737

> And voilà, we get the 5 NGC objects that are the most cited in literature, are not cluster

Reviewer 2: ```suggestion
And voilà, we get the 5 NGC objects that are the most cited in literature, are not clusters
```

           mesSpT                                                Collection of spectral types.
         journals                             Description of all used journals in the database

To join tables, any columns sharing the same name are possible links between tables.
To find the other possible joins, the `~astroquery.simbad.SimbadClass.list_linked_tables` method
can be useful. Here we look for possible links with the ``mesDiameter`` table

.. doctest-remote-data::

    >>> from astroquery.simbad import Simbad
    >>> Simbad.list_linked_tables("mesDiameter")
    <Table length=1>
     from_table from_column target_table target_column
       object      object      object        object
    ----------- ----------- ------------ -------------
    mesDiameter      oidref        basic           oid

The output indicates that the ``mesDiameter`` table can be linked to ``basic`` with the following
join statement: ``[...] mesDiameter JOIN basic ON mesDiameter.oidref = basic.oid [...]``.

.. graphviz:: simbad-er.gv
    :layout: neato
    :caption: A quick view of SIMBAD's tables. Hover the links to see the linked columns.

>     :alt: This interactive graph summarizes the information that can be obtained with `~astroquery.simbad.SimbadClass.list_tables` and `~astroquery.simbad.SimbadClass.list_linked_tables`.

Reviewer 2: are the links working?  I don't see hovers.  But this renders nicely!
<img width="727" alt="image" src="https://github.com/astropy/astroquery/assets/143715/94ceea6e-01b0-4803-818b-51f5f52518da">

Reviewer 2: is it possible to increase the font size though?  that might be a good idea, to match the overall fontsize of the document.  It's OK if that's not possible.

Author: Thanks for the review and all the English tips! The links work for me only when I use a local server to see the docs, there might be a bit of javascript to make it work in the sphinx-graphviz library, I'm not sure. I hope it will work too when it's live in astroquery's documentation. Otherwise it'll be a static graph and we'll have to change the legend. I'm looking for a way to change the font size :)

Author: A peek: 
![Peek 18-10-2023 15-45](https://github.com/astropy/astroquery/assets/16650466/1ec6ee03-2d07-4ef1-90b7-e0171839f7a7)

Author: The fontsize is changed... but in a bit of a ugly way. It has to be specified in each node in the DOT file and [it only works with pixels](https://graphviz.org/docs/attrs/fontsize/ "https://graphviz.org/docs/attrs/fontsize/"), not 'em'. I chose 16px as it seems to be the standard size of most browsers but it won't adapt to the readers' browsers parameters if they changed it.

For the hovering labels, I don't know. I tried a fresh python environment to generate the docs and it still worked. I noticed the tooltips only appear for me when I explicitly select the browser's window by clicking on it... Otherwise it might be an upstream bug with sphinx-graphviz?

Also, should I add in the comments of the `query_tap.rst` file that there is a script that can generate the file anew in case Simbad's schema changes? https://github.com/ManonMarchand/simbad-ER-diagram/blob/main/simbad-and-graphviz.py

      object      object      object   ... object        object       
    ---------- ----------- ----------- ... ------ --------------------
        biblio      biblio        TEXT ...        meta.record;meta.bib
        biblio      oidref      BIGINT ...         meta.record;meta.id
           ref    abstract UNICODECHAR ...                 meta.record
           ref     bibcode        CHAR ...            meta.bib.bibcode
           ref         doi     VARCHAR ...          meta.code;meta.bib
           ref     journal     VARCHAR ...            meta.bib.journal
           ref   last_page     INTEGER ...               meta.bib.page
           ref    nbobject     INTEGER ...                 meta.number
           ref      oidbib      BIGINT ...        meta.record;meta.bib
           ref        page     INTEGER ...               meta.bib.page
           ref       title        CLOB ...                  meta.title
           ref      volume     INTEGER ...             meta.bib.volume
           ref        year    SMALLINT ...          meta.note;meta.bib

`~astroquery.simbad.SimbadClass.list_columns` can also be called with a keyword argument.

> This returns columns from any table for witch the  given keyword is either in the table name,

Reviewer 2: ```suggestion
This returns columns from any table for witch the given keyword is either in the table name,
```
